inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoRenderer.java

```diff
@@ -108,35 +110,62 @@ public static PdfAnnoModel render(AnnotatorState aState, VDocument aVDoc, String
     private static Span convertToPdfAnnoSpan(VSpan aVspan, String aColor, String aDocumentText,
                                              PdfExtractFile aPdfExtractFile)
     {
-        VRange range = aVspan.getRanges().get(0); // TODO: handle multiple ranges
-        // use offset pre and post string to increase uniqueness of annotation text
-        int offset = 20;
-        int start = range.getBegin() <= offset ? 0 : range.getBegin() - offset;
-        int end = range.getEnd() < aDocumentText.length() - offset
-            ? range.getEnd() + offset : aDocumentText.length() - 1;
-
-        String annotatedText = aDocumentText.substring(start, end);
-        String preText = annotatedText.substring(0,
-            (range.getBegin() <= offset ? range.getBegin() - start : offset));
-        String postText = annotatedText.substring(annotatedText.length() -
-            (range.getEnd() < aDocumentText.length() - offset
-                ? offset : aDocumentText.length() - range.getEnd()), annotatedText.length()) ;
-        // remove whitespaces as they are not present in PDFExtract text
-        String cleanAnnotatedText = annotatedText.replaceAll("\\s", "");
-
-        int index = aPdfExtractFile.getStringContent().indexOf(cleanAnnotatedText);
-        if (index < 0) {
-            LOG.error("Could not map exisiting annotation with id " + aVspan.getVid().toString());
-            return null;
-        } else {
-            start = index + preText.replaceAll("\\s", "").length() + 1;
-            end = index + cleanAnnotatedText.length() - postText.replaceAll("\\s", "").length();
-            PdfExtractLine first = aPdfExtractFile.getStringPdfExtractLine(start);
-            PdfExtractLine last = aPdfExtractFile.getStringPdfExtractLine(end);
-            return new Span(aVspan.getVid().getId(), first.getPage(), aColor,
-                cleanAnnotatedText.substring(start - index, end - index).replaceAll("\\s", ""),
-                first.getPosition(), last.getPosition());
+        int docTextLen = aDocumentText.length();
+        int vSpanBegin = aDocumentText.length();
+        int vSpanEnd = -1;
+        // search for begin of the first range and end of the last range and use those
+        for (VRange range : aVspan.getRanges())
+        {
+            if (range.getBegin() < vSpanBegin) vSpanBegin = range.getBegin();
+            if (range.getEnd() > vSpanEnd) vSpanEnd = range.getEnd();
         }
+
+        // use an offset before and after string to find a unique text snippet for an annotation
+        // begin with 0 and increase by 1 until a unique text is found
+        // offset is limited to length of a quarter of the text
+        for (int offset = 0; offset < docTextLen / 4 + 1; offset++)
```

Comment:
Why is the offset limited to 25% of the text?

Reply:
I thought it wouldn't be possible to find a unique mapping if there were two or more possible candidates found when offset reaches 25% pre and post annotated text however I just figured out negative examples.
It was a misconception of mine.
I will change the limit to document text length.

Reply:
Do you have unit tests for the matching algorithm?

Reply:
Not yet implemented.
That's something I wanted to talk to you but haven't done yet.
I wanted to write tests for the `render` method in `PdfAnnoRenderer` class.
I had a look at brat tests and wanted to do it similar however I encountered some problems which I was not able to resolve and therefore didn't include the tests for now.
Maybe we can discuss in gitter, but I need to have a look at this again.

Reply:
Ping me on Gitter when you wish to discuss.

Reply:
@reckart Should a unit test for the `render` method be added in this PR?

Reply:
Would be good.

Reply:
@reckart seems like the issue for the test disappeared.. don't know exactly why this is the case, however it is gone.
I copied over `tcf04-karin-wl.xml` from Webanno to test resources folder.
Is this ok or should I somehow import it from the webanno-brat package?

Reply:
You can copy it, no problem.